Update tool recommendations

jeanas opened the pull request:
This is my proposal to address #1468.

Fixes #1468

<!-- readthedocs-preview python-packaging-user-guide start -->
----
📚 Documentation preview 📚: https://python-packaging-user-guide--1474.org.readthedocs.build/en/1474/guides/tool-recommendations/

<!-- readthedocs-preview python-packaging-user-guide end -->

chrysle commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,140 @@
-If you're familiar with Python packaging and installation, and just want to know
-what tools are currently recommended, then here it is.
+The Python packaging landscape consists of many different tools. For many tasks,
+the Python Packaging Authority (PyPA, the umbrella organization which
```
```suggestion
the :term:`Python Packaging Authority <Python Packaging Authority (PyPA)>` (PyPA, the umbrella organization which
```

chrysle commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,140 @@
-=================================
+  If you use setuptools, please be aware that it contains many deprecated
+  features which are currently kept for compatibility, but should not be used.
+  For example, do not use ``python setup.py`` invocations
```
```suggestion
  For example, do **not** use ``python setup.py`` invocations
```

jeanas commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,140 @@
+:ref:`modernize-setup-py-project` for migration advice).
+
+
+Lock files
```
Yes, but follow-up PR?

chrysle replied:
Yup.

jeanas commented:
@chrysle Thanks for reviewing! I've addressed your comments, sometimes with slightly different wordings.

jeanas commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
+  If you use setuptools, please be aware that it contains many deprecated
+  features which are currently kept for compatibility, but should not be used.
+  For example, do **not** use ``python setup.py`` invocations
+  (cf. :ref:`setup-py-deprecated`), the ``python_requires`` argument to
```
```suggestion
  (cf. :ref:`setup-py-deprecated`), the ``setup_requires`` argument to
```

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
-what tools are currently recommended, then here it is.
+The Python packaging landscape consists of many different tools. For many tasks,
+the :term:`Python Packaging Authority <Python Packaging Authority (PyPA)>`
+(PyPA, the umbrella organization which encompasses many packaging tools and
```
I wouldn't say "organization", it's more like a bunch of people and projects that are sometimes connected loosely, and sometimes not.
https://www.pypa.io/ presents the PyPA as a working group, which is probably a better term to use here.

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
+the :term:`Python Packaging Authority <Python Packaging Authority (PyPA)>`
+(PyPA, the umbrella organization which encompasses many packaging tools and
+maintains this guide) purposefully does not make a blanket recommendation; for
+example, the reason there exist many build backends is that the landscape was
```
```suggestion
example, the reason there are many build backends is that the landscape was
```

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
+(PyPA, the umbrella organization which encompasses many packaging tools and
+maintains this guide) purposefully does not make a blanket recommendation; for
+example, the reason there exist many build backends is that the landscape was
+opened in order to enable the development of new backends serving certain users'
```
```suggestion
opened up in order to enable the development of new backends serving certain users'
```

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
+Package Index (PyPI)>`. It can install into the global environment or into
+virtual environments. You may want to read pip's recommendations for
+:doc:`secure installs <pip:topics/secure-installs>`. Pip is available by default
+in most Python installations.
```
".. through the stdlib package `ensurepip`".

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
+in most Python installations.
+
+Alternatively, for installing Python command line applications specifically,
+consider :ref:`pipx`, which is a wrapper around pip that installs each
```
"pip and virtualenv"?

chrysle replied:
> "pip and virtualenv"?

Pip and Venv, actually. Virtualenv might be supported in future versions.

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
+Alternatively, for installing Python command line applications specifically,
+consider :ref:`pipx`, which is a wrapper around pip that installs each
+application into a dedicated virtual environment. This avoids conflicts between
+the dependencies of different applications. On Linux, pipx is especially
```
"applications that are distributed as Python package distributions through the PyPI"?

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
+consider :ref:`pipx`, which is a wrapper around pip that installs each
+application into a dedicated virtual environment. This avoids conflicts between
+the dependencies of different applications. On Linux, pipx is especially
+important because it also avoids conflicts with the system (which are the reason
```
Not sure I'd separate Linux from others here — it's bad to sure globally shared envs in so many contexts. In GNU/Linux you can break some system software, including package managers, but a similar influence exists in other OS-independent situations.

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
 
-* Use `pip-tools`_, :ref:`pipenv`, or `poetry`_ to generate the fully-specified
-  application-specific dependencies, when developing Python applications.
+- Hatchling, which is part of :ref:`Hatch` (but can be used without
```
It's a part of the same Git repository but is published a fully separate distribution on PyPI. This should either clarify that it's a part of the Hatch ecosystem or just drop the implication.

Hatchling is a build backend, but Hatch is a frontend, installer, resolver, virtualenv manager, command runner, workflow tool. Let's not smash them together.

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
-.. _secure manner: https://pip.pypa.io/en/latest/topics/secure-installs/
-.. _pip-tools: https://github.com/jazzband/pip-tools
-.. _Poetry: https://python-poetry.org/
+- :ref:`setuptools`, the historical build backend. It can be configured
```
It's not historical. It's old, but it's also modern. It has a lot of legacy in its own codebase but that's an implementation detail.

cc @jaraco @abravalheri

jeanas replied:
I didn't mean that it shouldn't be used anymore, only that it was the only choice for a long time. I think this is useful to put on this page because many people will be wondering why the heck they didn't have to make any choice of backend when they created a package years ago. Maybe `` :ref:`setuptools`, which used to be the only build backend `` would be clearer?

webknjaz replied:
Yep, that would be better. Thanks!

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
-.. _pip-tools: https://github.com/jazzband/pip-tools
-.. _Poetry: https://python-poetry.org/
+- :ref:`setuptools`, the historical build backend. It can be configured
+  programmatically through the :file:`setup.py` file (but for basic metadata,
```
Or `setup.cfg`, if we're listing the configuration details here. But then, why don't we do this for the other list items? If we're listing the config files, why not mention `hatch.toml`, `.pip-tools.toml`, `Pipfile` above too?

I'd drop these from here too.

jeanas replied:
I'm just trying to list which backends support plugins / programmatic config. (I think this is important, because those that do have an ecosystem of plugins around them.) `pyproject.toml` was originally not mentioned here, but added per https://github.com/pypa/packaging.python.org/pull/1474#discussion_r1438610971

webknjaz replied:
Okay, but I still want https://github.com/pypa/packaging.python.org/pull/1474/files#r1466878549.

I'll mark this as resolved in favor of that thread.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
-  :term:`PyPI <Python Package Index (PyPI)>`. [1]_ [2]_ Depending on how :ref:`pip`
-  is installed, you may need to also install :ref:`wheel` to get the benefit
-  of wheel caching. [3]_
+- Flit-core, part of :ref:`Flit` (but usable standalone). It is meant to be a
```
Same as with hatch — let's not mix self-sufficient front ends into a section about backends.

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
-=============================
+If you have :term:`extension modules <extension module>` and want to distribute
+wheels for multiple platforms, use :ref:`cibuildwheel` as part of your CI setup
+to build distributable wheels.
```
Let's mention `pip wheel` that can be used as a low-level piece of automation for building platform-specific wheels in places where `cibuildwheel` may be unavailable. With a note that pypa/build can be used for the same.

jeanas replied:
I thought the pip developers were unhappy with `pip wheel` and considering to deprecate it (but I may be misremembering, and I can't find the pip issue right now). @pfmoore @pradyunsg What's your opinion?

pfmoore replied:
I'm not aware of plans to deprecate `pip wheel`, but it's worth noting that:

1. *Building* wheels isn't its sole/primary purpose - it downloads wheels, or downloads sdists and builds them into wheels. `pip wheel .` builds a wheel of the current project, but somewhat as a side issue (we could potentially use a cached wheel if we had a way to be sure it matched the source code).
2. There's no pip command to build a sdist, and there may well never be one.

If you want to build a local project, `build` is a much more appropriate tool.

jeanas replied:
Ok, then I'd rather not mention `pip wheel` here.

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
+=================
 
-----
+The standard tool for this task is :ref:`twine`.
```
Let's mention that some of the workflow tools (hatch, poetry, flit) invoke it under the hood.

Also, let's link the secretless publishing guide which wraps it but should probably be the first choice for any GitHub-hosted projects using GHA: https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/

jeanas replied:
> Let's mention that some of the workflow tools (hatch, poetry, flit) invoke it under the hood.

I don't think they do? I didn't find any trace of this in the Hatch, Flit and Poetry repos.

webknjaz replied:
Ah, interesting. I thought they did.

@jeanas still worth mention the GHA, I think.

jeanas replied:
Sorry, I forgot about this one, done.

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
-       recommended for users that need cross-version consistency.
+These are tools that combine many features in one command line application, such
+as automatically managing virtual environments for a project, building
+distributions, uploading to PyPI, or creating and using lock files.
```
```suggestion
distributions,
uploading to PyPI, or creating and using tool-specific lock files.
They often call the tools mentioned above under the hood.
```

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
-       set across different versions of Python, and (unlike ``distutils``),
-       recent versions of ``setuptools`` support all of the modern metadata
-       fields described in :ref:`core-metadata`.
+- :ref:`Hatch`,
+- :ref:`Flit`,
+- :ref:`PDM`,
+- :ref:`Poetry`.
```
I suppose `Pipenv` can be mentioned in this section too.

eli-schwartz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
-    on the scientific community.
+For packages with :term:`extension modules <extension module>`, you may use
+setuptools, but consider using a build system dedicated to the language the
+extension is written in, such as Meson or CMake for C/C++, or Cargo for Rust,
```
Just want to point out that meson and cmake aren't specific to C/C++

They also support C#, Cuda, Swift, Asm, ObjC, ObjC++, Fortran.
Meson supports D, rust, vala, Java, and cython. CMake supports Hip and ISPC.

They are particularly valuable in cases where you have multiple languages, e.g. a mixture of rust and C++ and Fortran in a single project.

Meson also explicitly allows you to select `*.py files for installation`, do code generation of .py files (and other files), compile data such as gettext .po -> .mo catalogs, produce bundled `*.exe`s, and in general any form of custom build step. In theory CMake allows you to do precisely the same thing, although the scikit-build(-core) docs don't clearly indicate whether they support packaging that into a wheel.

tl;dr saying to use it "to support a dedicated extension language" is vastly underselling them. Even the "C/C++" part is simply *factually* wrong, if you need C/C++ support then setuptools works okay. The killer feature in the data science community is support for *Fortran*.

webknjaz replied:
@eli-schwartz are you okay with how it's been updated after your comment?

eli-schwartz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,144 @@
 
-* Use :ref:`virtualenv` or :doc:`venv <python:library/venv>` to isolate
-  application-specific dependencies from a shared Python installation. [4]_
+Popular :term:`build backends <build backend>` for pure-Python packages include:
```
```suggestion
Popular :term:`build backends <build backend>` for pure-Python packages include (in alphabetical order):
```

... and sort them.

abravalheri commented on source/guides/tool-recommendations.rst:
```diff
@@ -79,23 +90,16 @@ Popular :term:`build backends <build backend>` for pure-Python packages include:
-setuptools, but consider using a build system dedicated to the language the
-extension is written in, such as Meson or CMake for C/C++, or Cargo for Rust,
-and bridging this build system to Python using a dedicated build backend:
+setuptools. However, you can also use a build system with dedicated support for
+the language the extension is written in. For example, you could choose Meson or
+CMake for C, C++, Fortran and many other compiled languages; or Cargo for Rust
+specifically. You can then bridge this build system to Python using a dedicated
+build backend:
```
This is probably more direct:

```suggestion
a build system with dedicated support for the language the extension is written in:

- :ref:`setuptools` - natively supports C/C++ (with 3rd-party plugins for Go and Rust);

- :ref:`meson-python` - C, C++, Fortran, and Rust and other languages supported by Meson;

- :ref:`scikit-build-core` - C, C++, Fortran and other languages supported by CMake;

- :ref:`maturin` - Rust, via Cargo.
```

abravalheri commented on source/guides/tool-recommendations.rst:
```diff
@@ -62,12 +63,22 @@ environment, for reproducibility purposes.
+- Poetry-core_ (developed with but separate from :ref:`Poetry`). It is extensible
+  through plugins.
+
+- :ref:`setuptools` (which used to be the only build backend). It can be configured
```
```suggestion
- :ref:`setuptools` (which used to be the only build backend). It can be configured
  using modern standards like `pyproject.toml`, but can also be extended
  and supports customisation via `setup.py`.
```

abravalheri commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,161 @@
 
-* If you're looking for management of fully integrated cross-platform software
-  stacks, consider:
+  If you use setuptools, please be aware that it contains many deprecated
+  features which are currently kept for compatibility, but should not be used.
```
```suggestion
  If you use setuptools, please be aware that some features that predate
  standardisation efforts are now deprecated and only *temporarily kept* for compatibility.
```

jeanas commented:
I believe I've addressed all review comments. There were many reviews and it's not clear to me whether reviewers are now OK with the PR or just didn't have the time to take another look.

webknjaz commented:
I didn't have time, but I hope to find some.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
+.. _nox: https://nox.thea.codes
+.. _pdm-backend: https://backend.pdm-project.org
+.. _poetry-core: https://pypi.org/project/poetry-core/
+.. _tox: https://tox.wiki
```
Tox is wired through intersphinx.

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
-.. _distribute: https://pypi.org/project/distribute
+.. _flit-core: https://pypi.org/project/flit-core/
+.. _hatchling: https://pypi.org/project/hatchling/
+.. _nox: https://nox.thea.codes
```
Nox is wired through intersphinx.

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
-       choice for packaging.
 
-.. _distribute: https://pypi.org/project/distribute
+.. _flit-core: https://pypi.org/project/flit-core/
```
Flit is wired through intersphinx.

jeanas replied:
Yes, but I'd like to point to Flit-core here, and preferably not to the main Flit docs, to emphasize the distinction. (The text is `` Flit-core_ -- developed with but separate from :ref:`Flit`. ``.)

webknjaz replied:
Yeah, I wasn't completely sure about this one. Though, I see that you still applied the suggested change where I opted to link the doc that talks about `[build-system]` which is closer to the middle ground, it seems.

Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
+These tools allow you to define and invoke "tasks", such as running tests,
+compiling documentation, regenerating some files, etc. Each task can be executed
+in a dedicated virtual environment with the dependencies it requires. The most
+common tools for this are tox_ and nox_.
```
```suggestion
common tools for this are :doc:`tox <tox:index>` and :doc:`nox <nox:index>`.
```

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
 
-Application dependency management
-=================================
+- Flit-core_ -- developed with but separate from :ref:`Flit`. A minimal and
```
```suggestion
- :doc:`Flit-core <flit:pyproject_toml>` -- developed with but separate from :ref:`Flit`. A minimal and
```

jeanas replied:
Ah, I only see this now. Well, ok, why not.

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
+===================
+
+:ref:`Pip` is the standard tool to install packages from :term:`PyPI <Python
+Package Index (PyPI)>`. It can install into the global environment or into
```
Also, "virtual" (as a concept/word) is not an opposite of "global". I think this may be confusing to some.

webknjaz replied:
How about "isolated" vs. "shared"?

webknjaz replied:
@jeanas @pfmoore ideas?

jeanas replied:
Maybe we should just remove this sentence? Or replace it with “It can install into arbitrary environments.”?

pfmoore replied:
I think remove the sentence. It isn't in the version of the text that's currently live, and I don't think it adds anything.

jeanas replied:
Done.

webknjaz replied:
Marking as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
+:doc:`ensurepip <python:library/ensurepip>`.
+
+Alternatively, consider :ref:`pipx` for the specific use case of installing Python
+command line applications that are distributed through PyPI.
```
I think it's wrong to emphasize CLI apps, while realistically, it is useful in the context of any packages that are considered distributing apps. GUI included, maybe even web(ish) apps with strict pins could be included. So I'd just drop this clarification as not factual.
```suggestion
applications that are distributed through PyPI.
```

webknjaz replied:
Alternatively, clarify that there's more to it:
```suggestion
command line (and other types of) applications that are distributed through PyPI.
```

webknjaz replied:
I guess that's debatable. But fine. I don't have a better suggestion anyway. Do that.

webknjaz replied:
Marking as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
+command line applications that are distributed through PyPI.
+Pipx is a wrapper around pip and venv that installs each
+application into a dedicated virtual environment. This avoids conflicts between
+the dependencies of different applications, and also with the system
+(especially on Linux).
```
```suggestion
the dependencies of different applications, and also with the
ones installed system-wide making use of the same Python
interpreter (especially on GNU/Linux).
```

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
+
+Build backends
+==============
 
```
I feel like this might be the place to remind the readers not to treat this list as sorted by popularity.

```suggestion

.. important::

   Please, remember, this document does not seek to steer the reader towards
   a one-stop show solution. It merely enumerates tools observed within the
   packaging community. Different use-cases often need specialized workflows.

```

jeanas replied:
I'm not sure I understand your meaning of “one-stop shop”. I'm not a native speaker, but I'd read it as “a tool which does everything”, not “a build backend that you should always use”. Also I feel there is already a lot of warning at the beginning of the page.

webknjaz replied:
True. Let's wait for more opinions, then. And you're right, I did mean "a tool that does everything". Merely because a lot of people demand one without understanding that it entails. It seems to me that build backends are treated similarly even though, it's a different context. So I kinda wanted to assert the position once again in this section.

webknjaz replied:
@pfmoore do you have any thoughts on this?

pfmoore replied:
Looks ok to me. I’d say “a particular solution” rather than “a one-stop shop solution”. But otherwise it’s good.

jeanas replied:
Applied, with a slightly more concise wording.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
+point to some tools that are widely recognized, and also makes some
+recommendations of tools that you should *not* use because they are deprecated
+or insecure.
+
```
Should we mention that some tools focus on many things?
```suggestion

.. tip::

   Note that while most tools are is only mentioned under one
   section in this document, some of them cover several use
   cases.

```

jeanas replied:
Well, the intent is that all tools presented are single-purpose, except those which are presented under “Integrated workflow tools” (but maybe that title should be more explicit?).

webknjaz replied:
Perhaps yes. I also don't fully understand why tox/nox are ripped off of that section. I'd say they belong in it just as well.

jeanas replied:
I've merged them.

webknjaz replied:
Marking as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
+  If you use setuptools, please be aware that some features that predate
+  standardisation efforts are now deprecated and only *temporarily kept*
+  for compatibility.
 
-* Use :ref:`virtualenv` or :doc:`venv <python:library/venv>` to isolate
-  project-specific dependencies from a shared Python installation. [4]_
+  In particular, do **not** use direct ``python setup.py`` invocations. On the
+  other hand, configuring setuptools with a :file:`setup.py` file is still fully
+  supported, although it is recommended to use the modern :ref:`[project] table
+  in pyproject.toml <writing-pyproject-toml>` whenever possible and keep
+  :file:`setup.py` only if programmatic configuration is needed. See
+  :ref:`setup-py-deprecated`.
 
-* If you're looking for management of fully integrated cross-platform software
-  stacks, consider:
+  Other examples of deprecated features you should **not** use include the
+  ``setup_requires`` argument to ``setup()`` (use the :ref:`[build-system] table
+  <pyproject-guide-build-system-table>` in :file:`pyproject.toml` instead), and
+  the ``easy_install`` command (cf. :ref:`pip vs easy_install`).
```
How about wrapping this with `.. caution::`?

webknjaz replied:
Marking this as resolved.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,168 @@
-       :ref:`setuptools`), e.g. if you need to install from :term:`Eggs <Egg>`
-       (which pip doesn't support).  For a detailed breakdown, see :ref:`pip vs
-       easy_install`.
+Task runners
+============
```
I think it's more accurate to say
```suggestion
Workflow tools
==============
```

Saying "task runners" reduce them to a GNU/make replacement. But really, there's things like venv management that this skips.

jeanas replied:
I agree that this could be more precise, but if we make this change then I don't know how to call the section about Flit, Hatch, PDM, Pipenv and Poetry, which is currently called “Integrated workflow tools”. Thoughts?

pfmoore replied:
Fundamentally, my thought here is that we're trying too hard to classify the tools available in the ecosystem. We might be better leaving it to the tools themselves to choose how they want to present themselves, and not have any headings at all...

jeanas replied:
I'm not enthusiastic. The problem that this page solves is “it's too hard to understand how the tools fit together”. If it doesn't classify them, then it doesn't serve its purpose well, IMO.

webknjaz replied:
I suppose, I'd be in favor of combining these two sections into one. I feel like there's not enough differences in general to justify outing tox/nox vs. poetry/hatch et al.

jeanas replied:
It looks like not everyone has the same mental model as me on the nox/tox vs. Flit/Hatch/PDM/Pipenv/Poetry distinction.

For background, the two I've used the most are tox and Hatch, and I've sometimes used nox (in this repo). Tox is an environment manager that lets me define virtual environments and run tasks in them. Meanwhile, Hatch also does environment management and task execution, but also: building, uploading to PyPI, and managing Python interpreters (plus reformatting the code). So I view tox as a single-purpose tool (environment manager) along with other single-purpose tools (build, twine, pyenv) and Hatch as a tool that combines the features of lots of single-purpose tools (build, twine, tox, pyenv).

Merging these two sections is possible but I'm not sure how I'd fill the description.

Making the list task-oriented is also possible, but then Flit, Hatch, Poetry and PDM all are going to get listed in three categories (building, publishing, managing environments, except Flit doesn't manage environments).

This also makes me think that by my categorization, Pipenv should be with tox and nox. (For my defense, I know very little about it except that it sparked controversies years ago.)

webknjaz replied:
Technically, tox is used in the same way. People automate calling the underlying tools (like build or twine) and have tox as a workflow tool with a unified interfaces. It doesn't have some things built into the codebase but it also has plugins and stuff. So the use case for it is exactly the same, it just allows more low-level configuration. And it does implement some things that others don't (like parallel execution of several things), but that's just one of the differences. Other tools also differ between each other.

All of these are workflow tools. They just allow different levels of configuration and different have things implemented by default. But from the project contributor (end-user of those tools / the caller), they provide a CLI to call those processes as a part of their workflow.

@gaborbernat correct me if I'm missing something...

gaborbernat replied:
You're correct.

jeanas replied:
Ok, I've merged the two sections.

webknjaz replied:
Marking as resolved, then.

webknjaz commented:
@jeanas another round of review complete.

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,167 @@
+     In particular, do **not** use direct ``python setup.py`` invocations. On the
+     other hand, configuring setuptools with a :file:`setup.py` file is still fully
+     supported, although it is recommended to use the modern :ref:`[project] table
+     in pyproject.toml <writing-pyproject-toml>` whenever possible and keep
```
I'd like this to avoid ignoring the existence of `setup.cfg`:
```suggestion
     in pyproject.toml <writing-pyproject-toml>` (or :file:`setup.cfg`) whenever possible and keep
```

webknjaz commented on source/guides/tool-recommendations.rst:
```diff
@@ -4,111 +4,172 @@
 
-Publishing platform migration
-=============================
+For projects hosted on GitHub, PyPI recommends using :ref:`trusted publishing
```
Maybe, mention it's the PyPA?
```suggestion
For projects hosted on GitHub, the PyPA and the PyPI recommend using :ref:`trusted publishing
```

webknjaz replied:
Alternatively,
```suggestion
For projects hosted on GitHub, it is recommended to use the :ref:`trusted publishing
```

jeanas replied:
I'll go for the second, it's simpler.

webknjaz reviewed (approved):
I marked all the threads as resolved as there doesn't seem to be anything more actionable or blocking.

webknjaz commented:
The CI seems to be failing for an unrelated reason:
```console
(tutorials/installing-packages: line  444) broken    https://stackoverflow.com/a/14638025 - 403 Client Error: Forbidden for url: https://stackoverflow.com/a/14638025
```

The deps don't seem to differ compared to `main` so it's probably SO that's blocking HTTP requests from GH somehow. Will need to figure it out first.

willingc commented:
@webknjaz Maybe for CI, we change the SO link in the source to: https://stackoverflow.com/questions/14637979/how-to-permanently-set-path-on-linux-unix/14638025#14638025 It's where the CI link actually redirects if entered in the browser.

webknjaz commented:
@willingc that might work but needs checking. It should be done in a separate PR, though. An alternative would be adding a pattern to `linkcheck_ignore` if it's found that this problem exists with any StackOverflow links.

webknjaz commented:
@jeanas while that's being sorted out, what do you think about possibly linking https://pip.pypa.io/en/latest/topics/workflow/ / https://github.com/pypa/pip/pull/12503 in follow-ups? Not sure if it'd go in this specific doc, though.
